internal/cache.go

```diff
@@ -0,0 +1,204 @@
+// Copyright 2021 VMware, Inc.
+// SPDX-License-Identifier: BSD-2-Clause
+
+package internal
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"io/fs"
+	"os"
+	"path/filepath"
+
+	v1 "github.com/google/go-containerregistry/pkg/v1"
+)
+
+// cachedImage wraps a remote image with a local layer cache at dir
+type cachedImage struct {
+	v1.Image
+	dir string
+}
+
+// cachedLayer wraps a remote layer with a local cache at dir
+type cachedLayer struct {
+	v1.Layer
+	dir string
+}
+
+// teeLayerDump allows to download a layer while saving it in a cached file
+// at the same time.
+// Once the teeLayerDump is closed and if the stream has been fully read,
+// the local cached file can be used instead of the remote download.
+type teeLayerDump struct {
+	dump       io.Reader
+	layer      *cachedLayer
+	f          *os.File
+	rc         io.ReadCloser
+	compressed bool
+	done       bool
+}
+
+// NewCachedImage wrap a v1.Image, usually a remote one, so that layer downloads
+// are cached at the given dir.
+// Once the first download happens from an image layer, the following download
+// won't happen and instead the layer is copied from that local directory cache.
+func NewCachedImage(img v1.Image, dir string) v1.Image {
+	return &cachedImage{Image: img, dir: dir}
+}
+
+// wrapAsCachedLayer returns a cachedLayer ensuring the wrapping only happens
+// once, as it will not wrap a layer that is already a cachedLayer
+//
+// The reason for this is that a v1.Image interface can return layers from
+// different methods and we do not know if one of them consumes layers from
+// another so we do not know when the wrap must happen. We need to do it on
+// every method but avoid wrapping the layer several times.
+func (img *cachedImage) cachedLayer(layer v1.Layer) cachedLayer {
+	if ly, ok := (layer).(cachedLayer); ok {
+		return ly // Already a cached layer, so just return it
+	}
+	return newCachedLayer(layer, img.dir)
+}
+
+// Layers implements v1.Image's Layers wrapping layers as cachedLayers
+func (img *cachedImage) Layers() ([]v1.Layer, error) {
+	layers, err := img.Image.Layers()
+	for i, layer := range layers {
+		layers[i] = img.cachedLayer(layer)
+	}
+	return layers, err
+}
+
+// LayerByDigest implements v1.Image's LayerByDigest wrapping layers as cachedLayers
+func (img *cachedImage) LayerByDigest(hash v1.Hash) (v1.Layer, error) {
```

Comment:
are you sure this is the easiest way to achieve this? It seems like a lot of work to re-implement (or even call internally to the original implementation) to implement a cache layer.

isn't it possible to just download an image in a directory and load it from there instead?

Reply:
Not to mention that now we will need to make sure that if the interface changes we need to update our implementation for what in my mind is a side requirement.

Reply:
> isn't it possible to just download an image in a directory and load it from there instead?
Possible yes, inefficient and uncool also. The first download of each layer will be done to a temporary file completely before it can be used to again read the file completely and pass it to the end consumer.

Also that is not very easy to do anyways, without doing something VERY similar to what we have here. Keep in mind that this is being used from within the **MultiRefWrite** call:
~~~go
		refToImage[change.ImageReference] = internal.NewCachedImage(change.Image, cacheDir)
		logger.Printf("Processing image %s\n", change.ImageReference.Name())
	}

	logger.Printf("Writing %d images...\n", len(refToImage))
	if err := tarball.MultiRefWrite(refToImage, imagesFile); err != nil {
		return "", err
	}
~~~

You would have to pass images down to that method anyways, so you would have to replicate the file locally, not just the layers but everything...

Also....

> Not to mention that now we will need to make sure that if the interface changes we need to update our implementation for what in my mind is a side requirement.

If `v1.Image` or `v1.Hash` or `v1.Layer` interfaces, which are public, change this would be the least of our problems. O do not think those are subject to change. They are interfaces to interact with images that are local, remote, backed by a tarball, or something else. And even if you would want to avoid it... the call still takes a map of v1.Image so... how are you going to be decoupled anyway?

Reply:
At the end of the day **images are what is passed to the tarball writer**, images have layers and layers are the things we care to cache, the rest is cheap to keep remote or at least replicating the image locally completely might not be as easy or easier that what we have here, while being also less efficient and having more latency.